analyzer/analyzer.go

```diff
@@ -1,26 +1,19 @@
 package analyzer
 
 import (
-	"context"
-	"encoding/json"
-	"sort"
+	"os"
+	"strings"
+	"sync"
 
-	v1 "github.com/google/go-containerregistry/pkg/v1"
+	godeptypes "github.com/aquasecurity/go-dep-parser/pkg/types"
 	"golang.org/x/xerrors"
 
-	"github.com/aquasecurity/fanal/cache"
-	"github.com/aquasecurity/fanal/extractor"
-	"github.com/aquasecurity/fanal/extractor/docker"
 	"github.com/aquasecurity/fanal/types"
-	godeptypes "github.com/aquasecurity/go-dep-parser/pkg/types"
 )
 
 var (
-	osAnalyzers      []OSAnalyzer
-	pkgAnalyzers     []PkgAnalyzer
-	libAnalyzers     []LibraryAnalyzer
-	commandAnalyzers []CommandAnalyzer
-	additionalFiles  []string
+	analyzers       []analyzer
+	configAnalyzers []configAnalyzer
```

Comment:
Design: Since we are modifying this I would like to propose that we should eliminate global package state if we can. I think we can achieve this by exposing a function that can take in a argument of `[]analyzer` and the user of the library can pass in the various analyzers they would like to register.

More commentary here: https://dave.cheney.net/2017/06/11/go-without-package-scoped-variables

Reply:
Yes, I usually avoid package scoped variables, but this is `Registration`pattern and many go standard packages use it.
https://sourcegraph.com/github.com/golang/go/-/blob/src/database/sql/sql.go#L35

The blog author doesn't like the `Registration` pattern, but I feel there are some advantages. It makes introducing plugin easy. A build tag can toggle whether we use the analyzer or not.

```
// +build commercial

package myanalyzer

func init() {
    analyzer.Register(myAnalyzer{})
}

...
```

```
$ go build // myanalyzer is not used
$ go build -tags commercial  // myanalyzer is used
```

Also, Go has `plugin` package.
https://golang.org/pkg/plugin/

```
// +build commercial

package myanalyzer

func init() {
    analyzer.Register(myAnalyzer{})
}

...
```

```
$ go build -buildmode=plugin -o myanalyzer.so myanalyzer.go
```

```
func main() {
    files, err := ioutil.ReadDir("plugin/")
    if err != nil {
        panic(err)
    }

    for _, file := range files {
        // init() is called
        p, err := plugin.Open(file.Name())
        if err != nil {
            fmt.Fprintln(os.Stderr, err)
            os.Exit(-1)
        }
    }
}
...
```

```
$ go run main.go // myanalyze is not used
$ mv myanalyzer.so plugin/
$ go run main.go // myanalyzer is used
```

I think Go maintainers also know a package scoped variable is not good, but they selected extensibility.